src/Controller/HelperController.php

```diff
@@ -63,11 +64,21 @@ class HelperController
      */
     protected $validator;
 
+    /**
+     * @var DataTransformerResolver
+     */
+    private $resolver;
+
     /**
      * @param ValidatorInterface $validator
      */
-    public function __construct(Environment $twig, Pool $pool, AdminHelper $helper, $validator)
-    {
+    public function __construct(
+        Environment $twig,
+        Pool $pool,
+        AdminHelper $helper,
+        $validator,
+        DataTransformerResolver $resolver
```

Comment:
This is a BC break

Reply:
Is logical. You [asked](https://github.com/sonata-project/SonataAdminBundle/pull/5937#discussion_r390256695) to create a new service. The new service needs to be injected as a dependency into the constructor. Changing the constructor interface break BC.
If we did not add a new service, then there would be no problems with BC. I can now do without a service and make a separate PR to add service for a major release.

Reply:
Does
```suggestion
       ?DataTransformerResolver $resolver = null
```
avoid the BC break @core23 @peter-gribanov ?

With something like
```
if (!($resolver instanceof DataTransformerResolver)) {
   @trigger_error(....)
   $resolver = new DataTransformerResolver();
} else {
    $this->resolver = $resolver;
}
```

Reply:
@VincentLanglet Looks good. Any idea which message to use in `trigger_error()`?

Reply:
WDYT about something like:

Calling HelperController::construct without a DataTransformerResolver as fourth argument is deprecated since 3.x and will be throw an error in 4.0.

Reply:
I add this hook